src/v/archival/archival_policy.cc

```diff
@@ -219,6 +237,136 @@ static ss::output_stream<char> make_null_output_stream() {
     return ostr;
 }
 
+ss::future<offset_to_file_pos_result> convert_begin_offset_to_file_pos(
+  model::offset begin_inclusive,
+  const ss::lw_shared_ptr<storage::segment>& segment,
+  model::timestamp base_timestamp,
+  ss::io_priority_class io_priority) {
+    auto ix_begin = segment->index().find_nearest(begin_inclusive);
+    size_t scan_from = ix_begin ? ix_begin->filepos : 0;
+    model::offset sto = ix_begin ? ix_begin->offset
+                                 : segment->offsets().base_offset;
+    auto reader_handle = co_await segment->reader().data_stream(
+      scan_from, io_priority);
+    auto ostr = make_null_output_stream();
+
+    size_t bytes_to_skip = 0;
+    model::timestamp ts = base_timestamp;
+    auto res = co_await storage::transform_stream(
+      reader_handle.take_stream(),
+      std::move(ostr),
+      [begin_inclusive, &sto, &ts](model::record_batch_header& hdr) {
+          if (hdr.last_offset() < begin_inclusive) {
+              // The current record batch is accepted and will contribute to
+              // skipped length. This means that if we will read segment
+              // file starting from the 'scan_from' + 'res' we will be
+              // looking at the next record batch. We might not see the
+              // offset that we're looking for in this segment. This is why
+              // we need to update 'sto' per batch.
+              sto = hdr.last_offset() + model::offset(1);
+              // TODO: update base_timestamp
```

Comment:
If this TODO is still relevant, can it turn into a github issue?

Reply:
I have created https://github.com/redpanda-data/redpanda/issues/6990